lib/srv/sshserver.go

```diff
@@ -538,71 +552,109 @@ func (s *Server) HandleNewChan(nc net.Conn, sconn *ssh.ServerConn, nch ssh.NewCh
 			log.Errorf("failed to parse request data: %v, err: %v", string(nch.ExtraData()), err)
 			nch.Reject(ssh.UnknownChannelType, "failed to parse direct-tcpip request")
 		}
-		sshCh, _, err := nch.Accept()
+		ch, _, err := nch.Accept()
 		if err != nil {
 			log.Infof("could not accept channel (%s)", err)
 		}
-		go s.handleDirectTCPIPRequest(sconn, sshCh, req)
+		go s.handleDirectTCPIPRequest(sconn, ch, req)
 	default:
 		nch.Reject(ssh.UnknownChannelType, fmt.Sprintf("unknown channel type: %v", channelType))
 	}
 }
 
+// handleDirectTCPIPRequest does the port forwarding
 func (s *Server) handleDirectTCPIPRequest(sconn *ssh.ServerConn, ch ssh.Channel, req *sshutils.DirectTCPIPReq) {
-	// ctx holds the session context and all associated resources
+	// ctx holds the connection context and keeps track of the associated resources
 	ctx := newCtx(s, sconn)
 	ctx.isTestStub = s.isTestStub
 	ctx.addCloser(ch)
+	defer ctx.Infof("direct-tcp closed")
 	defer ctx.Close()
 
-	ctx.Infof("opened direct-tcpip channel: %#v", req)
 	addr := fmt.Sprintf("%v:%d", req.Host, req.Port)
-	ctx.Infof("connecting to %v", addr)
+	ctx.Infof("direct-tcpip channel: %#v to --> %v", req, addr)
 	conn, err := net.Dial("tcp", addr)
 	if err != nil {
 		ctx.Infof("failed to connect to: %v, err: %v", addr, err)
 		return
 	}
 	defer conn.Close()
+	// audit event:
+	s.EmitAuditEvent(events.PortForwardEvent, events.EventFields{
+		events.PortForwardAddr: addr,
+		events.EventLogin:      ctx.login,
+		events.LocalAddr:       sconn.LocalAddr().String(),
+		events.RemoteAddr:      sconn.RemoteAddr().String(),
+	})
 	wg := &sync.WaitGroup{}
 	wg.Add(1)
 	go func() {
 		defer wg.Done()
-		written, err := io.Copy(ch, conn)
-		ctx.Infof("conn to channel copy closed, bytes transferred: %v, err: %v",
-			written, err)
+		io.Copy(ch, conn)
 		ch.Close()
 	}()
 	wg.Add(1)
 	go func() {
 		defer wg.Done()
-		written, err := io.Copy(conn, ch)
-		ctx.Infof("channel to conn copy closed, bytes transferred: %v, err: %v",
-			ctx, written, err)
+		io.Copy(conn, ch)
 		conn.Close()
 	}()
 	wg.Wait()
-	ctx.Infof("direct-tcp closed")
+}
+
+// handleTerminalResize is called by the web proxy via its SSH connection.
+// when a web browser connects to the web API, the web proxy asks us,
+// by creating this new SSH channel, to start injecting the terminal size
+// into every SSH write back to it.
+//
+// this is the only way to make web-based terminal UI not break apart
+// when window changes its size
+func (s *Server) handleTerminalResize(sconn *ssh.ServerConn, ch ssh.Channel) {
+	// the party may not be immediately available for this connection,
+	// keep asking for a full second:
+	for i := 0; i < 10; i++ {
+		party := s.reg.PartyForConnection(sconn)
+		if party == nil {
+			time.Sleep(time.Millisecond * 100)
+			continue
+		}
+		// this starts a loop which will keep updating the terminal
+		// size for every SSH write back to this connection
+		party.termSizePusher(ch)
```

Comment:
may be this logic can be simplified if `termiSizePusher` is a member of the session, not a party